File: ubel-node/src/go_runner.js
// GoModScanner.js
import fs   from "fs";
import path from "path";

/**
 * Scans Go modules projects for installed dependencies.
 *
 * Detection strategy:
 *   1.  A directory is a Go module root when it contains go.mod AND go.sum.
 *   2.  Installed packages are read from go.sum (the resolved/verified set)
 *       cross-referenced with go.mod for direct vs indirect classification.
 *   3.  PURL: pkg:golang/<module-path>@<version>
 *
 * go.mod format:
 *   module github.com/org/repo
 *   require (
 *     github.com/foo/bar v1.2.3
 *     github.com/baz/qux v0.4.0 // indirect
 *   )
 *
 * go.sum format:
 *   github.com/foo/bar v1.2.3 h1:<hash>=
 *   github.com/foo/bar v1.2.3/go.mod h1:<hash>=
 */
export class GoModScanner {

  static inventoryData = [];

  // ─────────────────────────────
  // PURL
  // ─────────────────────────────
  static _goPurl(modulePath, version) {
    // Strip leading "v" from semver tags for consistency; keep pseudo-versions as-is.
    const v = version ? version.replace(/^v/, "") : "";
    return `pkg:golang/${modulePath.toLowerCase()}@${v}`;
  }

  // ─────────────────────────────
  // Detect Go module root
  // ─────────────────────────────
  static _isGoRoot(dir) {
    return (
      fs.existsSync(path.join(dir, "go.mod")) &&
      fs.existsSync(path.join(dir, "go.sum"))
    );
  }

  // ─────────────────────────────
  // Parse go.mod
  // Returns {
  //   moduleName: string,
  //   requires: Map<lowercasePath, { path, version, indirect }>,
  //   replaces: Map<lowercasePath, { original, replacement, version }>
  // }
  // ─────────────────────────────
  static _parseGoMod(modPath) {
    const requires = new Map();
    const replaces = new Map();
    let moduleName = "";

    let content;
    try {
      content = fs.readFileSync(modPath, "utf8");
    } catch {
      return { moduleName, requires, replaces };
    }

    const lines = content.split("\n");
    let inRequire = false;
    let inReplace = false;

    for (const rawLine of lines) {
      const line = rawLine.trim();

      if (!line || line.startsWith("//")) continue;

      // module declaration
      if (line.startsWith("module ")) {
        moduleName = line.slice(7).trim().split(/\s/)[0];
        continue;
      }

      // Block openers
      if (line === "require (") { inRequire = true;  continue; }
      if (line === "replace (") { inReplace = true;  continue; }
      if (line === ")")          { inRequire = false; inReplace = false; continue; }

      // Inline single-line directives
      const singleRequire = line.match(/^require\s+(\S+)\s+(\S+)(\s*\/\/ indirect)?/);
      if (singleRequire) {
        const mp       = singleRequire[1];
        const ver      = singleRequire[2];
        const indirect = !!singleRequire[3];
        requires.set(mp.toLowerCase(), { path: mp, version: ver, indirect });
        continue;
      }

      if (inRequire) {
        // "github.com/foo/bar v1.2.3 // indirect"
        const m = line.match(/^(\S+)\s+(\S+)(\s*\/\/ indirect)?/);
        if (m) {
          const mp       = m[1];
          const ver      = m[2];
          const indirect = !!m[3];
          requires.set(mp.toLowerCase(), { path: mp, version: ver, indirect });
        }
        continue;
      }

      if (inReplace) {
        // "github.com/old/pkg => github.com/new/pkg v1.0.0"
        const m = line.match(/^(\S+)(?:\s+\S+)?\s+=>\s+(\S+)\s+(\S+)/);
        if (m) {
          replaces.set(m[1].toLowerCase(), {
            original:    m[1],
            replacement: m[2],
            version:     m[3]
          });
        }
        continue;
      }
    }

    return { moduleName, requires, replaces };
  }

  // ─────────────────────────────
  // Parse go.sum → Set of "<module>@<version>"
  // Only /go.mod lines give version presence; h1: lines confirm download.
  // We want the source-code hash lines (not /go.mod lines) to get the
  // actually-used modules.
  // ─────────────────────────────
  static _parseGoSum(sumPath) {
    const installed = new Map();   // lowercase path → { path, version }

    let content;
    try {
      content = fs.readFileSync(sumPath, "utf8");
    } catch {
      return installed;
    }

    for (const rawLine of content.split("\n")) {
      const line = rawLine.trim();
      if (!line) continue;

      const parts = line.split(" ");
      if (parts.length < 2) continue;

      const [modVer] = parts;
      // Skip go.mod-only entries
      if (modVer.endsWith("/go.mod")) continue;

      const atIdx  = modVer.lastIndexOf("@");
      if (atIdx < 0) continue;

      const modPath = modVer.slice(0, atIdx);
      const version = modVer.slice(atIdx + 1);
      const key     = modPath.toLowerCase();

      if (!installed.has(key)) {
        installed.set(key, { path: modPath, version });
      }
    }

    return installed;
  }

  // ─────────────────────────────
  // Scan a single Go module root
  // ─────────────────────────────
  static _scanProject(projectRoot) {
    const { moduleName, requires, replaces } = this._parseGoMod(
      path.join(projectRoot, "go.mod")
    );

    const sumEntries = this._parseGoSum(path.join(projectRoot, "go.sum"));

    if (!sumEntries.size && !requires.size) return [];

    // Build full index: prefer go.sum versions (actual downloaded); fall back
    // to go.mod requires (for local replace directives that don't appear in go.sum)
    const index = new Map();   // lowercase path → { path, version, indirect }

    for (const [key, { path: mp, version }] of sumEntries.entries()) {
      const req = requires.get(key);
      index.set(key, {
        path:     mp,
        version,
        indirect: req ? req.indirect : true
      });
    }

    // Include require-only entries (local modules / replace targets not in go.sum)
    for (const [key, { path: mp, version, indirect }] of requires.entries()) {
      if (!index.has(key)) {
        index.set(key, { path: mp, version, indirect });
      }
    }

    // Apply replace directives – update resolved path/version
    for (const [origKey, { replacement, version }] of replaces.entries()) {
      const entry = index.get(origKey);
      if (entry) {
        entry.path    = replacement;
        entry.version = version;
      }
    }

    const components = [];

    for (const [, { path: mp, version, indirect }] of index.entries()) {
      const id = this._goPurl(mp, version);

      components.push({
        id,
        name:         mp.toLowerCase(),
        version:      version.replace(/^v/, ""),
        type:         "library",
        license:      "unknown",
        ecosystem:    "golang",
        state:        "undetermined",
        scopes:       [],
        dependencies: [],   // go.sum doesn't encode the dep graph; populated below
        paths:        [projectRoot],
        project_root: projectRoot,
        _indirect:    indirect
      });
    }

    return components;
  }

  // ─────────────────────────────
  // Assign scopes
  // Go doesn't have a native dev-dep concept in go.mod; the only reliable
  // signal is whether a module is "indirect" (transitive) or direct.
  // UBEL maps: direct → prod, indirect → prod (transitive)
  // Test-only packages (ending in /testing, /testutil, etc.) → dev heuristic.
  // ─────────────────────────────
  static _assignScopes(inventory) {
    for (const comp of inventory) {
      if (!Array.isArray(comp.scopes)) comp.scopes = [];
      if (comp.scopes.length > 0) continue;

      const isTestHeuristic =
        comp.name.includes("/testing") ||
        comp.name.includes("/testutil") ||
        comp.name.includes("/mock")     ||
        comp.name.endsWith("test")      ||
        comp.name.includes("gomock")    ||
        comp.name.includes("testify");

      comp.scopes.push(isTestHeuristic ? "dev" : "prod");
    }
  }

  // ─────────────────────────────
  // Merge duplicates by PURL
  // ─────────────────────────────
  static mergeInventoryByPurl(components) {
    const map = new Map();

    for (const comp of components) {
      if (!map.has(comp.id)) {
        map.set(comp.id, { ...comp, paths: [...comp.paths] });
        continue;
      }
      const existing = map.get(comp.id);
      for (const p of comp.paths)  { if (!existing.paths.includes(p))  existing.paths.push(p); }
      for (const s of comp.scopes) { if (!existing.scopes.includes(s)) existing.scopes.push(s); }
    }

    return [...map.values()];
  }

  // ─────────────────────────────
  // ENTRY
  // ─────────────────────────────
  static async getInstalled(startDir = process.cwd()) {
    this.inventoryData = [];

    const visited = new Set();
    const raw     = [];

    function walk(dir) {
      let entries;
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (["node_modules", ".git", ".ubel", "vendor"].includes(entry.name)) continue;

        const full = path.join(dir, entry.name);

        if (GoModScanner._isGoRoot(full)) {
          const key = path.resolve(full);
          if (!visited.has(key)) {
            visited.add(key);
            raw.push(...GoModScanner._scanProject(full));
          }
          // Descend – Go workspaces (go.work) can nest multiple modules
        }

        walk(full);
      }
    }

    if (GoModScanner._isGoRoot(startDir)) {
      const key = path.resolve(startDir);
      if (!visited.has(key)) {
        visited.add(key);
        raw.push(...GoModScanner._scanProject(startDir));
      }
    }

    walk(startDir);

    const merged = this.mergeInventoryByPurl(raw);
    this._assignScopes(merged);

    for (const c of merged) delete c._indirect;

    this.inventoryData = merged;
    return merged.map(c => c.id);
  }
}

export default GoModScanner;

File: python/ubel/go_runner.py
"""
go_scanner.py — Go / go.mod scanner (getInstalled only).

Mirrors the JS GoModScanner:
  - Detects Go module roots (go.mod + go.sum)
  - Reads go.sum for resolved deps, go.mod for direct/indirect classification
  - Applies replace directives
  - Assigns scopes via test-path heuristics (no native dev-dep concept in Go)
  - Merges duplicate PURLs
  - PURL: pkg:golang/<module-path>@<version>  (leading "v" stripped)
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Set, Tuple


class GoModScanner:


    # ------------------------------------------------------------------ #
    # PURL                                                                 #
    # ------------------------------------------------------------------ #

    
    def _go_purl(self,module_path: str, version: str) -> str:
        v = version.lstrip("v") if version else ""
        return f"pkg:golang/{module_path.lower()}@{v}"

    # ------------------------------------------------------------------ #
    # Detect Go module root                                                #
    # ------------------------------------------------------------------ #

    
    def _is_go_root(self,directory: str) -> bool:
        return (
            os.path.exists(os.path.join(directory, "go.mod")) and
            os.path.exists(os.path.join(directory, "go.sum"))
        )

    # ------------------------------------------------------------------ #
    # Parse go.mod                                                         #
    # ------------------------------------------------------------------ #

    
    def _parse_go_mod(self,mod_path: str) -> Tuple[str, Dict[str, Any], Dict[str, Any]]:
        """
        Returns (module_name, requires, replaces).
        requires: lowercase_path → { path, version, indirect }
        replaces: lowercase_path → { original, replacement, version }
        """
        requires: Dict[str, Any] = {}
        replaces: Dict[str, Any] = {}
        module_name = ""

        try:
            content = Path(mod_path).read_text(encoding="utf-8")
        except OSError:
            return module_name, requires, replaces

        in_require = False
        in_replace = False

        for raw_line in content.splitlines():
            line = raw_line.strip()
            if not line or line.startswith("//"):
                continue

            if line.startswith("module "):
                module_name = line[7:].strip().split()[0]
                continue

            if line == "require (":
                in_require = True;  continue
            if line == "replace (":
                in_replace = True;  continue
            if line == ")":
                in_require = False; in_replace = False; continue

            # Inline single-line require
            m = re.match(r"^require\s+(\S+)\s+(\S+)(.*//\s*indirect)?", line)
            if m:
                mp, ver, ind = m.group(1), m.group(2), bool(m.group(3))
                requires[mp.lower()] = {"path": mp, "version": ver, "indirect": ind}
                continue

            if in_require:
                m = re.match(r"^(\S+)\s+(\S+)(.*//\s*indirect)?", line)
                if m:
                    mp, ver, ind = m.group(1), m.group(2), bool(m.group(3))
                    requires[mp.lower()] = {"path": mp, "version": ver, "indirect": ind}
                continue

            if in_replace:
                # "github.com/old/pkg => github.com/new/pkg v1.0.0"
                m = re.match(r"^(\S+)(?:\s+\S+)?\s+=>\s+(\S+)\s+(\S+)", line)
                if m:
                    replaces[m.group(1).lower()] = {
                        "original":    m.group(1),
                        "replacement": m.group(2),
                        "version":     m.group(3),
                    }
                continue

        return module_name, requires, replaces

    # ------------------------------------------------------------------ #
    # Parse go.sum                                                         #
    # ------------------------------------------------------------------ #

    
    def _parse_go_sum(self,sum_path: str) -> Dict[str, Any]:
        """Returns lowercase_path → { path, version }"""
        installed: Dict[str, Any] = {}

        try:
            content = Path(sum_path).read_text(encoding="utf-8")
        except OSError:
            return installed

        for raw_line in content.splitlines():
            line = raw_line.strip()
            if not line:
                continue
            parts = line.split()
            if len(parts) < 2:
                continue
            mod_ver = parts[0]
            if mod_ver.endswith("/go.mod"):
                continue
            at_idx = mod_ver.rfind("@")
            if at_idx < 0:
                continue
            mod_path = mod_ver[:at_idx]
            version  = mod_ver[at_idx + 1:]
            key = mod_path.lower()
            if key not in installed:
                installed[key] = {"path": mod_path, "version": version}

        return installed

    # ------------------------------------------------------------------ #
    # Scan one Go module root                                              #
    # ------------------------------------------------------------------ #

    
    def _scan_project(self,project_root: str) -> List[Dict[str, Any]]:
        _, requires, replaces = self._parse_go_mod(
            os.path.join(project_root, "go.mod")
        )
        sum_entries = self._parse_go_sum(
            os.path.join(project_root, "go.sum")
        )

        if not sum_entries and not requires:
            return []

        # Prefer go.sum versions (actually downloaded); fallback to go.mod
        index: Dict[str, Any] = {}
        for key, entry in sum_entries.items():
            req = requires.get(key)
            index[key] = {
                "path":     entry["path"],
                "version":  entry["version"],
                "indirect": req["indirect"] if req else True,
            }
        for key, entry in requires.items():
            if key not in index:
                index[key] = {
                    "path":     entry["path"],
                    "version":  entry["version"],
                    "indirect": entry["indirect"],
                }

        # Apply replace directives
        for orig_key, rep in replaces.items():
            if orig_key in index:
                index[orig_key]["path"]    = rep["replacement"]
                index[orig_key]["version"] = rep["version"]

        components: List[Dict[str, Any]] = []
        for entry in index.values():
            mp      = entry["path"]
            version = entry["version"]
            cid     = self._go_purl(mp, version)

            components.append({
                "id":           cid,
                "name":         mp.lower(),
                "version":      version.lstrip("v"),
                "type":         "library",
                "license":      "unknown",
                "ecosystem":    "golang",
                "state":        "undetermined",
                "scopes":       [],
                "dependencies": [],
                "paths":        [project_root],
                "project_root": project_root,
                "_indirect":    entry["indirect"],
            })

        return components

    # ------------------------------------------------------------------ #
    # Assign scopes                                                        #
    # ------------------------------------------------------------------ #

    
    def _assign_scopes(inventory: List[Dict[str, Any]]) -> None:
        test_patterns = [
            "/testing", "/testutil", "/mock", "test", "gomock", "testify"
        ]
        for comp in inventory:
            if not isinstance(comp.get("scopes"), list):
                comp["scopes"] = []
            if comp["scopes"]:
                continue
            name = comp["name"]
            is_test = any(p in name for p in test_patterns)
            comp["scopes"].append("dev" if is_test else "prod")

    # ------------------------------------------------------------------ #
    # Merge by PURL                                                        #
    # ------------------------------------------------------------------ #

    
    def merge_inventory_by_purl(components: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        merged: Dict[str, Dict] = {}
        for comp in components:
            cid = comp["id"]
            if cid not in merged:
                clone = dict(comp)
                clone["paths"]  = list(comp.get("paths", []))
                clone["scopes"] = list(comp.get("scopes", []))
                merged[cid] = clone
                continue
            existing = merged[cid]
            for p in comp.get("paths", []):
                if p not in existing["paths"]:
                    existing["paths"].append(p)
            for s in comp.get("scopes", []):
                if s not in existing["scopes"]:
                    existing["scopes"].append(s)
        return list(merged.values())

    # ------------------------------------------------------------------ #
    # ENTRY                                                                #
    # ------------------------------------------------------------------ #

    def get_installed(self, start_dir: str = ".") -> List[str]:
        self.inventory_data = []
        start_dir = os.path.abspath(start_dir)

        visited: Set[str] = set()
        raw: List[Dict[str, Any]] = []

        skip = {"node_modules", ".git", ".ubel", "vendor"}

        def walk(directory: str) -> None:
            try:
                entries = os.scandir(directory)
            except OSError:
                return
            with entries as it:
                for entry in it:
                    if not entry.is_dir(follow_symlinks=False):
                        continue
                    if entry.name in skip:
                        continue
                    full = entry.path
                    if self._is_go_root(full):
                        key = os.path.realpath(full)
                        if key not in visited:
                            visited.add(key)
                            raw.extend(self._scan_project(full))
                    walk(full)

        if self._is_go_root(start_dir):
            key = os.path.realpath(start_dir)
            if key not in visited:
                visited.add(key)
                raw.extend(self._scan_project(start_dir))

        walk(start_dir)

        merged = self.merge_inventory_by_purl(raw)
        self._assign_scopes(merged)

        for c in merged:
            c.pop("_indirect", None)

        self.inventory_data = merged
        return [c["id"] for c in merged]